packages/block-editor/src/components/block-list/block.js

```diff
@@ -197,6 +198,14 @@ function BlockListBlock( {
 		);
 	}
 
+	if ( Wrapper ) {
+		blockEdit = (
+			<Wrapper className="block-editor-block-list-block__wrapper">
```

Comment:
Could we allow a callback instead? This seems much more future proof, e.g. if you'd like to append something or have multiple wrappers.

Reply:
Would it still be possible to apply the `block-editor-block-list-block__wrapper` CSS class if I used a callback? I'm using that in `packages/block-editor/src/components/use-block-drop-zone/index.js` to determine drop zones for drag-and-drop.

Reply:
In `getNearestBlockIndex`, do we need to check that `element` is a block? What else could it be?

Reply:
I spoke to @talldan about this and he suggested:

1. Remove [the `wp-block` check in `getNearestBlockIndex`](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/use-block-drop-zone/index.js#L53-L56) entirely.
2. Replace [this line which sets `elements`](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/use-block-drop-zone/index.js#L329) to search for `wp-block` instead of naively looking at the container's children. Care must be taken to only select `wp-block`s that  one level deep, though. Something like: `const blockElements = Array.from( element.current.querySelectorAll( '.wp-block:not(.wp-block .wp-block)' ) )`.

Be sure to test in multiple browsers since we're using a relatively advanced query in `querySelectorAll`.

Another option is to use `useSelect` to fetch the children's client IDs and turn them into DOM nodes using the [`BlockNodes` context](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-list/root-container.js#L83). This requires a bit more wrangling, though.

Reply:
> Something like: `const blockElements = Array.from( element.current.querySelectorAll( '.wp-block:not(.wp-block .wp-block)' ) )`.

Testing this in DevTools, it doesn't look like using `:not()` in this way works. Maybe something this then:

```js
const blockElements = difference(
	Array.from( element.current.querySelectorAll( '.wp-block' ) ),
	Array.from( element.current.querySelectorAll( '.wp-block .wp-block' ) )
);
```

You get the idea! 😅

Reply:
On looking again, it seems like `BlockNodes` only stores selected blocks, so using `querySelectorAll` makes sense:
https://github.com/WordPress/gutenberg/blob/65258a077e4899ee4933c79077108f56f8b5ce6d/packages/block-editor/src/components/block-list/block-wrapper.js#L87-L102

Reply:
@noisysocks Alright, I'm currently working on implementing this using the `_.difference` solution you suggested.

I do wonder if perhaps there's a more performance-friendly way to get the elements. Maybe something like this? (I have no idea if it's faster or slower.)

```js
function getBlockElements( node ) {
	const list = [];

	for ( const childNode of node ) {
		if ( childNode.classList.contains( 'wp-block' ) ) {
			list.push( childNode );
		} else if ( childNode.children.length > 0 ) {
			const grandchildren = getBlockElements( childNode );
			if ( grandchildren.length > 0 ) {
				list.push( ...grandchildren );
			}
		}
	}

	return list;
}

const blockElements = getBlockElements( element.current );
```

Reply:
@ZebulanStanphill An option for optimizing it that I've tested on https://github.com/WordPress/gutenberg/pull/23952 is to only get the list of elements when the user starts dragging, something like this:
```jsx
const blockElements = useRef();
const hasPosition = !! position;
useEffect( () => {
  if ( hasPosition ) {
    blockElements.current = difference(
	Array.from( element.current.querySelectorAll( '.wp-block' ) ),
	Array.from( element.current.querySelectorAll( ':scope .wp-block .wp-block' ) )
    );
  }
}, [ hasPosition ] );
```

It makes the assumption that the blocks won't change in any way after the user has started dragging, which I think is ok.

Though I personally wouldn't worry too much about optimizing unless you see performance issues.